src/EditorFeatures/CSharpTest/StringIndentation/StringIndentationTests.cs

```diff
@@ -0,0 +1,660 @@
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.StringIndentation
+{
+    [UseExportProvider]
+    public class StringIndentationTests
+    {
+        private static async Task TestAsync(string contents)
+        {
+            using var workspace = TestWorkspace.CreateWorkspace(
+                TestWorkspace.CreateWorkspaceElement(LanguageNames.CSharp,
+                    files: new[] { contents.Replace("|", " ") },
+                    isMarkup: false));
+            var document = workspace.CurrentSolution.GetRequiredDocument(workspace.Documents.First().Id);
+            var root = await document.GetRequiredSyntaxRootAsync(default);
+
+            var service = document.GetRequiredLanguageService<IStringIndentationService>();
+            var regions = await service.GetStringIndentationRegionsAsync(document, root.FullSpan, CancellationToken.None).ConfigureAwait(false);
+
+            var actual = ApplyRegions(contents.Replace("|", " "), regions);
+            Assert.Equal(contents, actual);
+        }
+
+        private static string ApplyRegions(string val, ImmutableArray<StringIndentationRegion> regions)
+        {
+            var text = SourceText.From(val);
+            using var _ = ArrayBuilder<TextChange>.GetInstance(out var changes);
+
+            foreach (var region in regions)
+            {
+                var firstLine = text.Lines.GetLineFromPosition(region.IndentSpan.Start);
+                var lastLine = text.Lines.GetLineFromPosition(region.IndentSpan.End);
+                var offset = region.IndentSpan.End - lastLine.Start;
+
+                for (var i = firstLine.LineNumber + 1; i < lastLine.LineNumber; i++)
+                {
+                    var lineStart = text.Lines[i].Start;
+                    if (region.OrderedHoleSpans.Any(s => s.Contains(lineStart)))
+                        continue;
+
+                    changes.Add(new TextChange(new TextSpan(lineStart + offset - 1, 1), "|"));
+                }
+            }
+
+            var changedText = text.WithChanges(changes);
+            return changedText.ToString();
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.StringIndentation)]
+        public async Task TestEmptyFile()
+            => await TestAsync(string.Empty);
+
+        [Fact, Trait(Traits.Feature, Traits.Features.StringIndentation)]
+        public async Task TestLiteralError1()
+        {
+            await TestAsync(@"class C
+{
+    void M()
+    {
+        // not enough lines in literal
+        var v = """"""
+                """""";
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.StringIndentation)]
+        public async Task TestLiteralError2()
+        {
+            await TestAsync(@"class C
+{
+    void M()
+    {
+        // invalid literal
+        var v = """"""
+            text too early
+                """""";
+    }
```

Comment:
I don't think I saw a test for (and I'm curious what the feature does for):
```
    var v = """"""
               this is
       text too early
               fine
               """"""";
```

(Wrote this comment in CodeFlow, so spacing might be b0rked.. apologies)

Reply:
That string is reported (syntactically) as an error by the compiler.  Any strings in error get no tagger-line. Thanks!

Reply:
That answers my question, thanks. Was wondering how things would work in that case.